### ISSUE TITLE
[WIP] As a shopper, I can access a specific product thanks to a direct link

### ISSUE THREAD

**Issue by RachL** [openfoodfoundation/openfoodnetwork#12638]
**When working on this use clokify code #10879 - Product Modal Links**

## Description
**- As a:** potential shopper
**- On page:** new dedicate URL per shop
**- I want to be able to do:** 

I can access a direct link to a product.

## Design links

They can be found here: https://www.figma.com/proto/v2tzhoSc7U5pcdurI2tRcD/Engineering-Delivery---Front-Shop?page-id=1%3A9&node-id=1-43&viewport=548%2C492%2C0.1&t=SYSOQYUK0Nbn3Uh5-1&scaling=min-zoom&starting-point-node-id=1%3A43

Small change from this prototype: in case of several OC we would keep the current behavior of going straight to the shop after selecting the OC in the dropdown.

## Acceptance Criteria & Tests

**these questions need to be answer before starting developing this feature:** 

- where do other products appear in the shop once the shopper ask to see the full store?
- does the product open on a modal in mobile as well or a new page?

1. if the product is not linked to an open OC (with or without stock) or the visitor does not have access to the OC, this is the display they land on (**TODO: check with Mario that we have the shopfront closed home page when the shop is closed**):

![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/6525576/9bfcdd58-b5e7-496b-a520-e9c3677adbbf)

2.  If the product becomes out of stock but there is an open OC, a message states the product is unavailable:
![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/6525576/813d25ed-b4fd-4453-8c50-9d01299ef1a9)

3a. When the shopper has access to an open OC and there is stock, this OC is pre-selected in the blue dropdown and a "shopfor this product appears"

![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/6525576/dac0d8a0-9787-46da-97fe-78c40bb5fb8d)

3b. When the shopper has access to several OC, the red dropdown appears; When selecting an OC, the shopper gets straight to step 4 (same behavior than on the shopfront)

![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/6525576/c4fff105-c507-4885-9cc8-fd5cd3f2bf1f)

4. When lading on the shop, the shopper sees the product and all its variants and can choose to add to cart but has also a button to view all products (step 5)

5. When viewing all products **we need to decide where we put the linked products**

![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/6525576/492bac50-afc7-4585-9bc2-7cbdef90fae8)

**Comment by RachL** [comment]
@mariocarabotta @mkllnk  I've updated the above issue, but before starting the following is perhaps missing:

**Breadcrumb behavior:**

- once the shopper wants to see the full store, they reach the root of the breadcrumb. Where do other products of that store appear? I see 3 options maybe there are more: 

1.the previously visited product remain at the top 
2. we anchor the product at its "usual" place in the shop list of products and bring the shopper there
". We do nothing, and simply display the store as we would do with direct link to the store. => this would be my preference for now, but happy to follow another decision

**Indexing all products?**
- Do we index products that haven't been sold for more than X time? => I would say no
- Do we index only products for shops that are publicly available? )> i would say no as well. I would ask for a login in that case.

What do you think?

**URL logic as a gem?**
- @mkllnk FYI you've previously mentioned you would like the URL logic to gem. See comment [here](https://github.com/openfoodfoundation/openfoodnetwork/issues/10879#issuecomment-2292406078). I will let you decide if it's still a good idea :)

**Comment by mariocarabotta** [comment]
> **Breadcrumb behavior:**
> 
I agree to :do nothing, and simply display the store as we would do with direct link to the store.

> **Indexing all products?**

Agree on both points. In terms of how far in the past, I am not sure, 6 months maybe? It kind of depends also on how often we would do that batch process/review of what to index.

**Comment by mariocarabotta** [comment]
hi folks, here's a few considerations after I've spent a bit more time on this. Any feedback welcome!
for context: concept designs above were based on some broader redesign, now I'm trying to find the right spot in between so that we can move towards that, without building too much now.

1. Can we keep the whole shop navigation for now? This would be handy to retain links and the OC picker
2. If 1 is YES, then I would not build the breadcrumbs for now. I would prefer to see where we get to with categories navigation first, and then update this page once we have more clarity on it

Here's a few examples based on the two points above for your consideration. This is not the full list of scenarios listed above, I will take care of that once  we decide on nav and breadcrumbs.

**Single variant - before selecting OC**
Note the enterprise is shown under the product name

<img width="1280" height="1291" alt="Image" src="https://github.com/user-attachments/assets/80025300-7afc-4c59-9c5b-b3ffbda6e5e0" />

**Single variant - after selecting OC**

<img width="1280" height="1292" alt="Image" src="https://github.com/user-attachments/assets/dfc41d77-c496-401f-b5a9-0108f95f6e0e" />

**Multi variant (different enterprises) - before selecting OC**

<img width="1280" height="1298" alt="Image" src="https://github.com/user-attachments/assets/a64d7d8d-38f6-4df9-9e62-f8980c0c60d6" />

**Multi variant  (different enterprises) - after selecting OC**
Note each enterprise is shown under the variant name

<img width="1280" height="1409" alt="Image" src="https://github.com/user-attachments/assets/f9e5cdd4-4173-49c5-b424-c6c500586ea1" />

**Mobile example - single variant after selecting OC** 
<img width="360" height="1696" alt="Image" src="https://github.com/user-attachments/assets/298fca55-d44a-4e9c-a953-f882e0cb6463" />

For full context, this is where I was heading a while ago for a redesign that includes the top nav and the OC picker. You might see how the current proposal above is still heading towards this concept. I think that there are some items in the roadmap that could enable that if we think it's worth it.

<img width="1280" height="1476" alt="Image" src="https://github.com/user-attachments/assets/99f1a20f-6370-4898-939f-5369cda2ffff" />

**Comment by RachL** [comment]
> Agree on both points. In terms of how far in the past, I am not sure, 6 months maybe? It kind of depends also on how often we would do that batch process/review of what to index.

hmmm i'm pretty sure we do sell currently products that are older than 6 months. do you mean If the product were not sold since 6 months, do not index them? 

Regarding the layout: I think it looks great! I like that we are not introducing too much here and reusing the existing. That should be easier to deliver. 
One tiny comment: I'm not sure it will be clear that you need to click on "shop" to go back to the shop - given the shop tab is already highlighted when showing the product (maybe we should not highlight a tab in that case?).

This reminds me that I need to add a test: we need to test this feature as part of the white label settings as well.

I will update the issue once @mkllnk has given his feedback as well.

**Comment by mariocarabotta** [comment]
> > Agree on both points. In terms of how far in the past, I am not sure, 6 months maybe? It kind of depends also on how often we would do that batch process/review of what to index.
> 
> hmmm i'm pretty sure we do sell currently products that are older than 6 months. do you mean If the product were not sold since 6 months, do not index them?
> 

I'm thinking about the last time they were added to an OC, not creation date. But that might completely off :)

> Regarding the layout: I think it looks great! I like that we are not introducing too much here and reusing the existing. That should be easier to deliver. One tiny comment: I'm not sure it will be clear that you need to click on "shop" to go back to the shop - given the shop tab is already highlighted when showing the product (maybe we should not highlight a tab in that case?).
> 

Good point. I think having a separate issue for the breadcrumb gives us enough flexibility to address this separately once we can play with the page in staging.

**Comment by mkllnk** [comment]
> you've previously mentioned you would like the URL logic to gem. See comment https://github.com/openfoodfoundation/openfoodnetwork/issues/10879#issuecomment-2292406078. I will let you decide if it's still a good idea :)

Yes, that's still a good idea. I will work on that.

> Do we index products that haven't been sold for more than X time? => I would say no

What do you mean by index? If someone has a link to a product then we should show it as long as it belongs to a public order cycle, even if it closed. It was public in the past and could have been archived on another platform as well. I would also show products that appear in upcoming order cycles. That will enable shops to give VIP customers a sneak preview. The OC selector could also show things like:

- Currently available in: [oc1, oc2, oc3]
- Currently unavailable.
- Available from 17. October 2026

If we want to put a time limit on it, I wouldn't choose a fixed time as each shop has very different rhythms. It would introduce a rule of the last three or six or twelve order cycles. That adjusts with the shop's pace. But as I said above, I wouldn't put a limit on it.

**Comment by RachL** [comment]
> Available from 17. October 2026

this is a nice idea, but I'm afraid it only works well when there is only one OC coming up. Unless we add this date and message in the following section and don't show the OC selector? The date would then be the closest OC opening date where the product is shown.

<img width="304" height="93" alt="Image" src="https://github.com/user-attachments/assets/dd0c4356-8330-4d2a-b3af-d17393e9ac1f" />

**Comment by RachL** [comment]
@mkllnk regarding the gem, perhaps we want to open a dedicated issue? I guess it can be merged on its own - correct?

**Comment by mkllnk** [comment]
> regarding the gem, perhaps we want to open a dedicated issue? I guess it can be merged on its own - correct?

It can be separate. I can start with the URL `/products/123` and later add `/products/123-pink-lady-apple` to it. It's an optional improvement.

**Comment by mkllnk** [comment]
> I'm afraid it only works well when there is only one OC coming up.

I don't understand. I would just select the next OC containing that product to announce the date. And you can't choose the OC yet because it hasn't opened yet. The picture you shared is too negative, I think. We don't want to put people off thinking that the product is not available. We want them to know that it will be available soon. In some cases it could even be tomorrow or within the next minutes. People can just wait in that case.

**Comment by RachL** [comment]
@mkllnk yes but in some cases it might be never as well. We need to handle the case where no future OC are created. Some users don't create their OC in advance or close during 6 months of the year. We can't guess if they are temporarily closed or definitively closed.

**Comment by mkllnk** [comment]
> in some cases it might be never as well. We need to handle the case where no future OC are created.

Ah, yes, of course. We need a message for that as well.

**Comment by mkllnk** [comment]
@RachL, we need to decide on the product URL structure. We discussed several options in the delivery circle but wanted your input as well. Here are examples for a real product:

```
/fat-beets-food-hub/33880-bread-wood-fired-sourdough
/fat-beets-food-hub/products/33880-bread-wood-fired-sourdough
/fat-beets-food-hub/shop/33880-bread-wood-fired-sourdough
```

@mariocarabotta, @rioug and I prefer the `products` path so far. It makes it clear that you are looking at a product and it follows standard REST conventions. That may not be as clear in countries with other languages than English. The first option is the only language-agnostic one. The last builds on the current shop URL but I'm a bit wary of locking us into putting everything under shop in the future. We may want to add other pages in the future. What do you think?

**Comment by RachL** [comment]
@mkllnk I agree with your assessment for `/shop/`. Regarding `/products/` I think that if that's the standard approach we need to go for it. 

But what is the standard actually? E.g. Should the whole list of products also appear under `/products/` in the shopfront? It's out of scope here, but maybe something to document, otherwise we won't remember why the individual product is following the standard and not the list.

**Comment by mkllnk** [comment]
> Should the whole list of products also appear under /products/ in the shopfront?

Yes. The Rails default would be `/enterprises/3/products/` but we skip the `enterprises` path to have shorter URLs and use a permalink instead of the id. And then we renamed products to shop which was probably a gut feel decision at the time. But also, in those old times, Spree still used the route to `products` and the `shop` path was avoiding any potential conflicts in the routing. That is all gone now and we can do what we want. I don't see any need to change the shop URL. But we could add a redirect from `/permalink/products/` to `/permalink/shop` in case someone edits the URL manually.

Summary:

- We settled on `/fat-beets-food-hub/products/33880-bread-wood-fired-sourdough` unless new perspectives arise.
- We redirect from products to shop to catch a common UX issue.
